keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1184 @@
+  pemTrustAnchors: "<... PEM DATA ...>"
+```
+
+As the set of trust anchors for my CA changes due to rotations and revocations,
+my controller updates the status of the ClusterTrustBundle object.
+
+I then update my client pod to add a `trustAnchors` volume:
+```diff
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: client
+  name: client
+spec:
+  containers:
+    name: main
+    image: my-image
+    volumeMounts:
++    - mountPath: /var/run/example-com-server-tls-trust-anchors
++      name: example-com-server-tls-trust-anchors
++      readOnly: true
+  volumes:
++  - name: example-com-server-tls-trust-anchors
++    projected:
++      sources:
++      - pemTrustAnchors:
++          clusterTrustBundleName: example.com--server-tls-live
++          path: ca_certificates.pem
+```
+
+When my client pod is scheduled onto a node, Kubelet writes the chosen trust
+anchor set into the container filesystem of my main container.  My main
+container doesn't start until `ca_certificates.pem` is present and non-empty.
+
+I ensure that my client application is capable of reading a set of trust anchors
+from a file containing PEM-wrapped DER-formatted X.509 certificates at the
+location I've specified
+(`/var/run/example-com-server-tls-trust-anchors/ca_certificates.pem`).
+
+I ensure that my client application notices changes to `ca_certificates.pem`
+(using inotify or periodic polling), and properly updates its TLS stack when
+that happens.
+
+#### Trust Anchor Configuration for Webhook Backends
+
+In my cluster, I operate a collection of validating and mutating admission
+webhooks that use a server TLS certificate issued from my company's private CA.
+Today, when my private CA introduces a new root as part of a scheduled rotation
+(or incident response), I need to track down and adjust all
+ValidatingWebhookConfiguration and MutatingWebhookConfiguration objects in my
+cluster that refer to my private CA, and update their
+`.webhooks[].clientConfig.caBundle` fields.
+
+ClusterTrustBundle objects let me centralize the management of my private CA's
+roots within my cluster.
+
+To use it, I create a new ClusterTrustBundle containing the current roots of my
+private CA.
+
+```yaml
+apiVersion: v1alpha
+kind: ClusterTrustBunddle
+metadata:
+  name: example-com-private-ca
+spec:
+  signerName: "" # This ClusterTrustBundle isn't associated with a signer, so I
+                 # leave this empty.
+  pemTrustAnchors: "<... PEM DATA ...>"
+```
+
+Then, I update each ValidatingWebhookConfiguration and
+MutatingWebhookConfiguration object that currently refer to my private CA,
+removing the CABundle field and adding a reference to the ClusterTrustBundle
+object I created.
+
+```diff
+apiVersion: admissionregistration/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: my-validating-webhook
+webhooks:
+  name: frobnicator.example.com
+  clientConfig:
+    url: https://webhooks.example.com/frobnicator
+-    caBundle: "< long base64 data>"
```

Comment:
Hmmm.  That's problematic for the rollback case where you enable the feature gate, configure the feature, and then disable the feature gate.

Should we introduce some sort of selector enum, or maybe say that trustAnchors takes precedence over caBundle if it is set (and the feature gate is enabled)?

Reply:
I've updated the doc so that ClusterTrustBundleName takes precedence over the other TLS fields if it is non-empty.  That way, users don't have to unset CABundle in order to try out ClusterTrustBundleName, and they can freely toggle the feature gate.